### DISCUSSION TITLE
Large codebase in version 9.1.5. How best to upgrade without regression?

### DISCUSSION THREAD

**Discussion by niamleeson** [vercel/next.js#17102] [Help] [answered]
Hello,

We have a fairly large codebase in the version 9.1.5. We did not have the time to upgrade NextJs versions as they came out and we are behind. Is there a way to upgrade to the new version a page at a time so that the newer pages that are being added can use the new Next.Js version?

Not sure if this is relevant or not, but we have an exposed custom server for some custom routing logic which wasn't possible in version 9.1.5.

**Comment by M1ck0** [comment, marked as answer]
You could use [this](https://nextjs.org/docs/advanced-features/codemods) to help you to start

You also might find a use for [multi zones](https://nextjs.org/docs/advanced-features/multi-zones)

I think that multi zones are your best bet since it will allow you to have 2 different apps that work as 1. And you can just transfer parts of the old app to the new one until you finish the whole project.

If someone has any other solution it would be cool to post it here

**Reply by niamleeson** [reply]
Thank you for your response! I didn't realize there was a codemod! Awesome! I'll eventually use that.

I do love the idea of multizones and I did read about it, but I'm still not sure how you can have multiple versions of nextjs. Would you be able to link me or show me an example?

**Reply by M1ck0** [reply]
[Here](https://github.com/vercel/next.js/tree/canary/examples/with-zones) you can see an example from the official repo.

As you can see in you have `/blog` and `/home` pages and they both have their own `package.json` (which means they both have different node_modules directories with different packages), and other config files (you can look at them like they are separate projects).

I suggest you clone that repo and deploy it on Vercel (there is a tutorial inside that repo), to get a feel on how it works.

You just need to configure proxies when deploying on your own servers (there are loads of data about that too online)

**Reply by niamleeson** [reply]
Ohhh I thought that example only worked with Vercel. It works with containerized deployment to right?

**Reply by M1ck0** [reply]
Yeah. With Vercel it is just easy to configure.
When deploying to your own servers you need to set up proxies and you won't be using that 'vercel.json' file since it is for Vercel only.

Here is text from [docs](https://nextjs.org/docs/advanced-features/multi-zones)


**_You can also configure a proxy server to route using a set of routes like the ones above, e.g deploy the blog app to https://blog.example.com and the home app to https://home.example.com and then add a proxy server for both apps in https://example.com._**

**Reply by niamleeson** [reply]
Ohhh I get it. Thank you!

**Reply by karthik-js** [reply]
@M1ck0 I would love to know how can i share auth data from zone1 app to zone2 app?
